### PR TITLE
Fix/6790 return accepted for rejected tx added to broadcast

### PR DESCRIPTION
**Pull request by pxyxyrus**
Fixes Closes Resolves #

Issue 6790
https://github.com/NethermindEth/nethermind/issues/6790

#### What types of changes does your code introduce?

- [x] Bugfix (a non-breaking change that fixes an issue)
- [ ] New feature (a non-breaking change that adds functionality)
- [ ] Breaking change (a change that causes existing functionality not to work as expected)
- [ ] Optimization
- [ ] Refactoring
- [ ] Documentation update
- [ ] Build-related changes
- [ ] Other: _Description_

#### Requires testing

- [x] Yes
- [ ] No

#### If yes, did you write tests?

- [ ] Yes
- [x] No

#### Notes on testing

The change broke several existing tests, so the changes made in the test code were to make the test pass.
However, reviews are needed.

### CODE REVIEW

File: src/Nethermind/Nethermind.TxPool/TxPool.cs
Comment (Demuirgos) [inline src/Nethermind/Nethermind.TxPool/TxPool.cs:457]: what happened to ``_broadcaster.Broadcast`` call?

Comment (pxyxyrus) [reply src/Nethermind/Nethermind.TxPool/TxPool.cs:457]: Since the if branch now has the opposite condition, if the `tx` is persistent (and not a blob tx), it would run to the end of the function (line 462~483).

`_broadcaster.Broadcast` will be run at line 477.

Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.TxPool/TxPool.cs:457]: Not sure if running through all the normal flow here is best choice as this is a special case, will wait for @marcindsobczak opinion.

**Review by marcindsobczak — CHANGES_REQUESTED** [review]
There are a few things which should be done before returning `Accepted`:
- check if tx was successfully added to broadcaster collection - I recommend to modify `Broadcast` to `TryBroadcast` and return `true` if successfully added
- check if tx is not blob type (we shouldn't add blob tx to broadcaster if we didn't accepted it to standard blob pool - because we are not keeping actual blobs in broadcaster)

Optionally (not must have, but good to have) write a test trying to add different types of txs to already full tx pool and not full persistent pool (so basically testing adding txs scenario with rejecting from standard pool and adding to broadcaster collection) - it should reject blob tx and accept all the other types

**Comment by marcindsobczak** [comment]
In `TxBroadcaster` it can be something like:
```
        public bool TryBroadcast(Transaction tx, bool isPersistent)
        {
            if (isPersistent)
            {
                return TryStartBroadcast(tx);
            }

            BroadcastOnce(tx);
            return true;
        }

        private bool TryStartBroadcast(Transaction tx)
        {
            if (tx.Hash is null)
            {
                return false;
            }

            bool wasInserted = _persistentTxs.TryInsert(tx.Hash, tx.SupportsBlobs ? new LightTransaction(tx) : tx);

            // broadcast local tx only if MaxFeePerGas is not lower than configurable percent of current base fee
            // (70% by default). Otherwise only add to persistent txs and broadcast when tx will be ready for inclusion
            if (wasInserted && tx.MaxFeePerGas >= _baseFeeThreshold || tx.IsFree())
            {
                NotifyPeersAboutLocalTx(tx);
            }

            return wasInserted;
        }
```
current form:
```
        public void Broadcast(Transaction tx, bool isPersistent)
        {
            if (isPersistent)
            {
                StartBroadcast(tx);
            }
            else
            {
                BroadcastOnce(tx);
            }
        }

        private void StartBroadcast(Transaction tx)
        {
            // broadcast local tx only if MaxFeePerGas is not lower than configurable percent of current base fee
            // (70% by default). Otherwise only add to persistent txs and broadcast when tx will be ready for inclusion
            if (tx.MaxFeePerGas >= _baseFeeThreshold || tx.IsFree())
            {
                NotifyPeersAboutLocalTx(tx);
            }

            if (tx.Hash is not null)
            {
                _persistentTxs.TryInsert(tx.Hash, tx.SupportsBlobs ? new LightTransaction(tx) : tx);
            }
        }
```

**Comment by marcindsobczak** [comment]
Or add to persistent broadcast only if fee is at least X percent of current base fee (maybe even better? @LukaszRozmej )
```
        private bool TryStartBroadcast(Transaction tx)
        {
            // add to persistent txs and broadcast local tx only if MaxFeePerGas is not lower than configurable
            // percent of current base fee (70% by default)
            if (tx.Hash is not null
                && _persistentTxs.TryInsert(tx.Hash, tx.SupportsBlobs ? new LightTransaction(tx) : tx)
                && tx.MaxFeePerGas >= _baseFeeThreshold || tx.IsFree())
            {
                NotifyPeersAboutLocalTx(tx);
                return true;
            }

            return false;
        }
```

**Review by pxyxyrus — COMMENTED** [review]
Even after implementing the suggested solution, the following tests don't pass.

```
should_add_underpaid_txs_to_full_TxPool_only_if_local(True)

should_increase_nonce_when_transaction_not_included_in_txPool_but_broadcasted

Should_not_add_underpaid_tx_even_if_lower_nonces_are_expensive(9,False)

Should_not_replace_better_txs_by_worse_ones

Should_not_replace_ready_txs_by_nonce_gap_ones
```

The first two tests were updated to expect `Accepted`.

I'm being more careful about updating the other three tests.

File: src/Nethermind/Nethermind.TxPool/TxBroadcaster.cs
Comment (marcindsobczak) [inline src/Nethermind/Nethermind.TxPool/TxBroadcaster.cs]: After this change, we will return `false` if tx was inserted, but is not close to base fee. It will be possible, that we return `false`, but tx will be included in the block in the future. We should decide on one of two options:

1. just return `txInserted` at the end of the method. Even if tx is not good enough to send to peers right now, but if we add it to persistent collection, we need to return `true`
2. move attempt to insert tx inside `if` statement checking price, so just `return _persistentTxs.TryInsert(tx.Hash, tx.SupportsBlobs ? new LightTransaction(tx) : tx);` inside `if` and `return false` at the end

It depends if we want to add local tx with fee lower than 70% of `baseFee` or reject it. I don't have a strong preference here, you can decide. If I would need to make this decision, I will probably go with option 1

File: src/Nethermind/Nethermind.TxPool/TxBroadcaster.cs
Comment (marcindsobczak) [inline src/Nethermind/Nethermind.TxPool/TxBroadcaster.cs:100]: Minor cosmetic: would be good to rename method, right now we are not calling it if we want to `broadcast` (void), but when we want to `try to broadcast` (bool return depending on success). So I recommend renaming it to `TryBroadcast` and `TryStartBroadcast`

File: src/Nethermind/Nethermind.TxPool/TxBroadcaster.cs
Comment (marcindsobczak) [inline src/Nethermind/Nethermind.TxPool/TxBroadcaster.cs:125]: We will reject now to add local txs if fee is lower than X percent (70% as default) of base fee, which is fine

File: src/Nethermind/Nethermind.TxPool/TxBroadcaster.cs
Comment (marcindsobczak) [inline src/Nethermind/Nethermind.TxPool/TxBroadcaster.cs:118]: Ahh but order is wrong, we are firstly trying to insert and then checking fee, so it is possible that we will add tx to persistent collection, but return false. I made this mistake in proposed code in previous review
```suggestion
            if (tx is not null
                && (tx.MaxFeePerGas >= _baseFeeThreshold || tx.IsFree())
                && _persistentTxs.TryInsert(tx.Hash, tx.SupportsBlobs ? new LightTransaction(tx) : tx))
```

Comment (marcindsobczak) [reply src/Nethermind/Nethermind.TxPool/TxBroadcaster.cs:118]: And there is one more case, thare is a possibility that we will insert tx and evict it just after, so we need logic similar to the one from TxPool. Modify
`_persistentTxs.TryInsert(tx.Hash, tx.SupportsBlobs ? new LightTransaction(tx) : tx)`
to be
`_persistentTxs.TryInsert(tx.Hash, tx.SupportsBlobs ? new LightTransaction(tx) : tx, out var removed)`
and then check if `tx.Hash` is the same as `removed.Hash`. If hashes are the same it means that we didn't actually added tx to persistent collection

It is because we have some capacity, but under the hood we are actually firstly adding new tx (and having capacity max + 1) and then removing the worst one. So if we add and remove the same tx, it means that it wasn't actually added :D

Comment (marcindsobczak) [reply src/Nethermind/Nethermind.TxPool/TxBroadcaster.cs:118]: and then tests should work as expected, this is the reason why tests are failing

File: src/Nethermind/Nethermind.TxPool/TxBroadcaster.cs
Comment (marcindsobczak) [reply src/Nethermind/Nethermind.TxPool/TxBroadcaster.cs:100]: Worth to consider before merging

**Review by marcindsobczak — APPROVED** [review]
Looks good, there are some minor cosmetics which might be improved (I left a few new comments), but generally mergable

File: src/Nethermind/Nethermind.TxPool.Test/TxPoolTests.cs
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.TxPool.Test/TxPoolTests.cs:1642]: Maybe better to have TestCaseSource and generate test cases for all enum items, by default with `true` and just have exception for blobs. In that case if we add new tx type we will by default have a test that expects `true` here.

Comment (pxyxyrus) [reply src/Nethermind/Nethermind.TxPool.Test/TxPoolTests.cs]: I've changed the test accordingly to check all the values of `TxType`.

File: src/Nethermind/Nethermind.TxPool.Test/TxPoolTests.cs
Comment (pxyxyrus) [reply src/Nethermind/Nethermind.TxPool.Test/TxPoolTests.cs]: applied

**Pull request merged**
